src/main/java/org/broadinstitute/hellbender/engine/FeatureDataSource.java

```diff
@@ -288,9 +288,21 @@ public static boolean isGenomicsDBPath(final String path) {
                 throw new UserException("GenomicsDB inputs can only be used to provide VariantContexts.", e);
             }
         } else {
-            final Path featurePath = IOUtils.getPath(featureInput.getFeaturePath());
-            IOUtils.assertFileIsReadable(featurePath);
-            final FeatureCodec<T, ?> codec = (FeatureCodec<T, ?>) FeatureManager.getCodecForFile(featurePath, targetFeatureType);
+            final Class<FeatureCodec<T, ?>> codecClass = featureInput.getFeatureCodecClass();
+            FeatureCodec<T, ?> codec;
+            if (codecClass == null) {
+                final Path featurePath = IOUtils.getPath(featureInput.getFeaturePath());
+                IOUtils.assertFileIsReadable(featurePath);
+                codec = (FeatureCodec<T, ?>) FeatureManager.getCodecForFile(featurePath, targetFeatureType);
+                featureInput.setFeatureCodecClass((Class<FeatureCodec<T, ?>>)codec.getClass());
+            } else {
+                try {
+                    codec = codecClass.newInstance();
+                }
+                catch ( InstantiationException | IllegalAccessException e ) {
+                    throw new GATKException("Unable to automatically instantiate codec " + codecClass.getName());
+                }
+            }
```

Comment:
Can you extract a helper method here that takes a `FeatureInput` and returns a codec?